### ISSUE TITLE
Does Istio not support wss?

### ISSUE THREAD

**Issue by win5923** [istio/istio#49765]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

If I don't include certificates within the MQTT broker, both the MQTT and WS protocols can communicate. However, after adding certificates, the MQTTS protocol still works, but the WSS protocol fails to establish a connection.

kiali:
![image](https://github.com/istio/istio/assets/56353753/1cc9fd08-c6f4-4f31-8134-98b8e9ddf2cb)

### Version

using `defualt` Istio configuration profile.

```Text
$ istioctl version
client version: 1.20.3
control plane version: 1.20.3
data plane version: 1.20.3 (3 proxies)
$ kubectl version
Client Version: v1.29.0
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.5
```

### Additional Information
```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: mosquitto-gateway
  namespace: mosquitto-prod
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 1883
      name: mqtt
      protocol: TCP
    hosts:
    - "*"
  - port:
      number: 9090
      name: websocket
      protocol: HTTP
    hosts:
    - "*"
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: mosquitto-tcp
  namespace: mosquitto-prod
spec:
  hosts:
  - "*"
  gateways:
  - mosquitto-gateway
  tcp:
  - match:
    - port: 1883
    route:
    - destination:
        host: mosquitto-v2015
        port:
          number: 1883
      weight: 100
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: mosquitto-ws
  namespace: mosquitto-prod
spec:
  hosts:
  - "*"
  gateways:
  - mosquitto-gateway
  http:
  - match:
    - port: 9090
    route:
    - destination:
        host: mosquitto-v2015
        port:
          number: 9090
      weight: 100      
```
ws port:
istioctl proxy-config listeners -n istio-system istio-ingressgateway-64c48998d7-wkz7c --port 9090 -o json
```
[
    {
        "name": "0.0.0.0_9090",
        "address": {
            "socketAddress": {
                "address": "0.0.0.0",
                "portValue": 9090
            }
        },
        "filterChains": [
            {
                "filters": [
                    {
                        "name": "envoy.filters.network.http_connection_manager",
                        "typedConfig": {
                            "@type": "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager",
                            "statPrefix": "outbound_0.0.0.0_9090",
                            "rds": {
                                "configSource": {
                                    "ads": {},
                                    "initialFetchTimeout": "0s",
                                    "resourceApiVersion": "V3"
                                },
                                "routeConfigName": "http.9090"
                            },
                            "httpFilters": [
                                {
                                    "name": "istio.metadata_exchange",
                                    "typedConfig": {
                                        "@type": "type.googleapis.com/udpa.type.v1.TypedStruct",
                                        "typeUrl": "type.googleapis.com/io.istio.http.peer_metadata.Config",
                                        "value": {
                                            "upstream_discovery": [
                                                {
                                                    "istio_headers": {}
                                                },
                                                {
                                                    "workload_discovery": {}
                                                }
                                            ],
                                            "upstream_propagation": [
                                                {
                                                    "istio_headers": {}
                                                }
                                            ]
                                        }
                                    }
                                },
                                {
                                    "name": "envoy.filters.http.grpc_stats",
                                    "typedConfig": {
                                        "@type": "type.googleapis.com/envoy.extensions.filters.http.grpc_stats.v3.FilterConfig",
                                        "emitFilterState": true,
                                        "statsForAllMethods": false
                                    }
                                },
                                {
                                    "name": "istio.alpn",
                                    "typedConfig": {
                                        "@type": "type.googleapis.com/istio.envoy.config.filter.http.alpn.v2alpha1.FilterConfig",
                                        "alpnOverride": [
                                            {
                                                "alpnOverride": [
                                                    "istio-http/1.0",
                                                    "istio",
                                                    "http/1.0"
                                                ]
                                            },
                                            {
                                                "upstreamProtocol": "HTTP11",
                                                "alpnOverride": [
                                                    "istio-http/1.1",
                                                    "istio",
                                                    "http/1.1"
                                                ]
                                            },
                                            {
                                                "upstreamProtocol": "HTTP2",
                                                "alpnOverride": [
                                                    "istio-h2",
                                                    "istio",
                                                    "h2"
                                                ]
                                            }
                                        ]
                                    }
                                },
                                {
                                    "name": "envoy.filters.http.fault",
                                    "typedConfig": {
                                        "@type": "type.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault"
                                    }
                                },
                                {
                                    "name": "envoy.filters.http.cors",
                                    "typedConfig": {
                                        "@type": "type.googleapis.com/envoy.extensions.filters.http.cors.v3.Cors"
                                    }
                                },
                                {
                                    "name": "istio.stats",
                                    "typedConfig": {
                                        "@type": "type.googleapis.com/stats.PluginConfig",
                                        "disableHostHeaderFallback": true
                                    }
                                },
                                {
                                    "name": "envoy.filters.http.router",
                                    "typedConfig": {
                                        "@type": "type.googleapis.com/envoy.extensions.filters.http.router.v3.Router"
                                    }
                                }
                            ],
                            "tracing": {
                                "clientSampling": {
                                    "value": 100
                                },
                                "randomSampling": {
                                    "value": 1
                                },
                                "overallSampling": {
                                    "value": 100
                                },
                                "customTags": [
                                    {
                                        "tag": "istio.authorization.dry_run.allow_policy.name",
                                        "metadata": {
                                            "kind": {
                                                "request": {}
                                            },
                                            "metadataKey": {
                                                "key": "envoy.filters.http.rbac",
                                                "path": [
                                                    {
                                                        "key": "istio_dry_run_allow_shadow_effective_policy_id"
                                                    }
                                                ]
                                            }
                                        }
                                    },
                                    {
                                        "tag": "istio.authorization.dry_run.allow_policy.result",
                                        "metadata": {
                                            "kind": {
                                                "request": {}
                                            },
                                            "metadataKey": {
                                                "key": "envoy.filters.http.rbac",
                                                "path": [
                                                    {
                                                        "key": "istio_dry_run_allow_shadow_engine_result"
                                                    }
                                                ]
                                            }
                                        }
                                    },
                                    {
                                        "tag": "istio.authorization.dry_run.deny_policy.name",
                                        "metadata": {
                                            "kind": {
                                                "request": {}
                                            },
                                            "metadataKey": {
                                                "key": "envoy.filters.http.rbac",
                                                "path": [
                                                    {
                                                        "key": "istio_dry_run_deny_shadow_effective_policy_id"
                                                    }
                                                ]
                                            }
                                        }
                                    },
                                    {
                                        "tag": "istio.authorization.dry_run.deny_policy.result",
                                        "metadata": {
                                            "kind": {
                                                "request": {}
                                            },
                                            "metadataKey": {
                                                "key": "envoy.filters.http.rbac",
                                                "path": [
                                                    {
                                                        "key": "istio_dry_run_deny_shadow_engine_result"
                                                    }
                                                ]
                                            }
                                        }
                                    },
                                    {
                                        "tag": "istio.canonical_revision",
                                        "literal": {
                                            "value": "latest"
                                        }
                                    },
                                    {
                                        "tag": "istio.canonical_service",
                                        "literal": {
                                            "value": "istio-ingressgateway"
                                        }
                                    },
                                    {
                                        "tag": "istio.mesh_id",
                                        "literal": {
                                            "value": "cluster.local"
                                        }
                                    },
                                    {
                                        "tag": "istio.namespace",
                                        "literal": {
                                            "value": "istio-system"
                                        }
                                    }
                                ]
                            },
                            "httpProtocolOptions": {},
                            "serverName": "istio-envoy",
                            "streamIdleTimeout": "0s",
                            "useRemoteAddress": true,
                            "forwardClientCertDetails": "SANITIZE_SET",
                            "setCurrentClientCertDetails": {
                                "subject": true,
                                "cert": true,
                                "dns": true,
                                "uri": true
                            },
                            "upgradeConfigs": [
                                {
                                    "upgradeType": "websocket"
                                }
                            ],
                            "normalizePath": true,
                            "pathWithEscapedSlashesAction": "KEEP_UNCHANGED",
                            "requestIdExtension": {
                                "typedConfig": {
                                    "@type": "type.googleapis.com/envoy.extensions.request_id.uuid.v3.UuidRequestIdConfig",
                                    "useRequestIdForTraceSampling": true
                                }
                            }
                        }
                    }
                ]
            }
        ],
        "listenerFiltersTimeout": "0s",
        "trafficDirection": "OUTBOUND"
    }
]
```

**Comment by zengyuxing007** [comment]
You can configure the wss port (9090) as a tcp protocol or as a https passthrough.

ref: https://istio.io/latest/docs/tasks/traffic-management/ingress/ingress-sni-passthrough/

**Comment by win5923** [comment]
Hi, thanks for your response. I changed the Gateway protocol from `HTTP` to `TCP` for wss and was able to connect successfully. I'm curious about the reason behind this. Initially, I configured it based on https://github.com/istio/istio/blob/master/samples/websockets/README.md, and using `HTTP` for ws also allowed me to connect. However, following your suggestion to change HTTP to TCP resulted in a successful connection.
